### PR TITLE
Share file watchers for metadata references

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Under the covers, we were already reusing the actual metadata, but not sharing VisualStudioMetadataReference meant that each project would create it's own copies of file watchers for all metadata references. These were (hopefully) being consolidated by the Visual Studio file watching service, but the overhead of doing so was still much larger than not doing it in the first place.

<details><summary>Ask Mode template</summary>

### Customer scenario

Opening a large solution isn't as fast as we'd like it to be.

### Bugs this fixes

[bug needs to be filed if needed]

### Workarounds, if any

None, other than make your solution smaller.

### Risk

Moderate: generally changes how the lifetime of some objects are. Change is mechanical though and uses established patterns in the codebase.

### Performance impact

Improved! Because of #28475, we were no longer eagerly subscribing to file watchers, but in the Roslyn.sln case we were still calling AdviseFileChanges 18,000 times on the background thread. Although the file change service in VS would be sharing those, that's still a non-trivial amount of time we're occupying in a single-threaded service that other things need during load as well. By reducing our load we're removing work happening on that background thread, allowing other work to get done first. This also means we're reusing timestamp reading, which was causing lots of disk I/O (or at least CPU time) as well.

### Is this a regression from a previous update?

Nope.

### Root cause analysis

Always been designed this way.

### How was the bug found?

Perf trace analysis.

</details>

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs [hunk 1]
```diff
@@ -74,6 +74,16 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
...
         private readonly Dictionary<VisualStudioMetadataReference, CancellationTokenSource> _donotAccessDirectlyChangedReferencesPendingUpdate
             = new Dictionary<VisualStudioMetadataReference, CancellationTokenSource>();
+
+        /// <summary>
+        /// Holds the "original" snapshot in the project that needs to be updated due a reference changing. Because we defer updating after a metadata
+        /// reference change, we need to rememeber the original before so we can do the removal/add correctly.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="OnImportChangedAfterDelay(System.Threading.Tasks.Task, object)" /> is what consumes this.</remarks>
+        private readonly Dictionary<VisualStudioMetadataReference, PortableExecutableReference> _originalSnapshotOfReferenceNeedingUpdate
```
Comment (tmat) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: >_originalSnapshotOfReferenceNeedingUpdate [](start = 96, length = 41)

It's not clear that this just adds another piece of information to the references queued in ChangedReferencesPendingUpdate. Perhaps merge these two dictionaries or add a comment.

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: @tmat Merged the dictionaries now. Want to re-review?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs [hunk 2]
```diff
@@ -308,7 +322,7 @@ public ProjectInfo CreateProjectInfoForCurrentState()
                     compilationOptions: this.CurrentCompilationOptions,
                     parseOptions: this.CurrentParseOptions,
                     documents: _documents.Values.Select(d => d.GetInitialState()),
-                    metadataReferences: _metadataReferences.Select(r => r.CurrentSnapshot),
+                    metadataReferences: _metadataReferences.Select(r => r.Target.CurrentSnapshot),
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Convince me that it's always safe to just access `Target` in this file without needing to check nulls?

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: ReferenceCountedDisposable is in one of two states:

1. It's not disposed, so Target is non-null.
2. It's disposed, and Target will throw ObjectDisposedException.

https://github.com/dotnet/roslyn/blob/63433fbc11c5efaeef3f451f8f31df3e76e5684a/src/Workspaces/Core/Portable/Utilities/ReferenceCountedDisposable.cs#L114

Comment (Pilchie) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Then how are we confident throughout this file that it's not disposed?

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: The weaselly answer is "as confident as we were before" -- the places where we were disposing before are unchanged.

The more concrete answer is we do disposal in two places:

1. The project itself being removed.
2. `RemoveMetadataReferenceCore`, which is the same helper that removes the entry from `metadataReferences` as well.

Comment (Pilchie) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Thanks!

**Comment by jinujoseph** [comment]
Approved by @ManishJayaswal for 15.9

**Comment by ManishJayaswal** [comment]
We need to quantify the perf improvements here before this can go into 15.9. So removing the approval tag.

**Comment by jinujoseph** [comment]
Ok discussed with Jason, we are going to see if we can quantify the perf improvement. Moving this for 16.0 may not make sense as this was targeted fix for 15

**Comment by jasonmalinowski** [comment]
So some numbers for this change: we remove 1.2 seconds of CPU time albeit from a background thread, and it looks like it was removing a few hundred megabytes of allocations during solution load. It'll also be removing approximately half of all the AdviseFileChange calls we do in Roslyn.sln, which itself is occupying some mutexes in Visual Studio. @sharwell any other numbers you wanted to add?

**Comment by jasonmalinowski** [comment]
To explain the close quick: at this point this has all been rewritten in 16.0, and we don't think it's worth the benefit at this point to take for 15.9 given any risk.

**Pull request closed without merging**